components/resizable/nz-resizable.directive.ts

```diff
@@ -0,0 +1,325 @@
+ * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
+ */
+
+import { Platform } from '@angular/cdk/platform';
+import {
+  AfterViewInit,
+  Directive,
+  ElementRef,
+  EventEmitter,
+  HostListener,
+  Input,
+  NgZone,
+  OnDestroy,
+  Output,
+  Renderer2
+} from '@angular/core';
+
+import { InputBoolean } from 'ng-zorro-antd/core';
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+
+import { NzResizableService } from './nz-resizable.service';
+import { NzResizeHandleMouseDownEvent } from './nz-resize-handle.component';
+
+export interface NzResizeEvent {
+  width?: number;
+  height?: number;
+  col?: number;
+  mouseEvent?: MouseEvent;
+}
+
+@Directive({
+  selector: '[nz-resizable]',
+  providers: [NzResizableService],
+  host: {
+    '[class.nz-resizable]': 'true',
+    '[class.nz-resizable-resizing]': 'resizing',
+    '(mouseenter)': 'onMouseenter()',
+    '(mouseleave)': 'onMouseleave()'
+  }
+})
+export class NzResizableDirective implements AfterViewInit, OnDestroy {
+  @Input() nzBounds: 'window' | 'parent' | ElementRef<HTMLElement> = 'parent';
+  @Input() nzMaxHeight: number;
+  @Input() nzMaxWidth: number;
+  @Input() nzMinHeight: number = 40;
+  @Input() nzMinWidth: number = 40;
+  @Input() nzGridColumnCount: number = -1;
+  @Input() nzMaxColumn: number = -1;
+  @Input() nzMinColumn: number = -1;
+  @Input() @InputBoolean() nzLockAspectRatio: boolean = false;
+  @Input() @InputBoolean() nzPreview: boolean = false;
+  @Output() readonly nzResize = new EventEmitter<NzResizeEvent>();
+  @Output() readonly nzResizeEnd = new EventEmitter<NzResizeEvent>();
+  @Output() readonly nzResizeStart = new EventEmitter<NzResizeEvent>();
+
+  resizing = false;
+  elRect: ClientRect | DOMRect;
+  currentHandleEvent: NzResizeHandleMouseDownEvent;
+  ghostElement: HTMLDivElement | null;
+  el: HTMLElement;
+  sizeCache: NzResizeEvent | null;
+  private destroy$ = new Subject<void>();
+
+  constructor(
+    private elementRef: ElementRef<HTMLElement>,
+    private renderer: Renderer2,
+    private ngZone: NgZone,
+    private nzResizableService: NzResizableService,
+    private platform: Platform
+  ) {
+    this.nzResizableService.handleMouseDown$.pipe(takeUntil(this.destroy$)).subscribe(event => {
+      this.resizing = true;
+      this.currentHandleEvent = event;
+      this.setCursor();
+      this.nzResizeStart.emit({
+        mouseEvent: event.mouseEvent
+      });
+      this.elRect = this.el.getBoundingClientRect();
+    });
+  }
+
+  @HostListener('document:mouseup', ['$event'])
+  onMouseup($event: MouseEvent): void {
+    this.ngZone.runOutsideAngular(() => {
+      if (this.resizing) {
+        this.resizing = false;
+        this.nzResizableService.documentMouseUp$.next();
+        this.endResize($event);
+      }
+    });
+  }
+
+  @HostListener('document:mousemove', ['$event'])
+  onMousemove($event: MouseEvent): void {
```

Comment:
IMHO, `mousemove` and `mouseup` handlers should only be attached only after users started dragging. Otherwise, we would bind a lot of event listeners. Refer to the implementation of CDK. https://github.com/angular/components/blob/master/src/cdk/drag-drop/drag-drop-registry.ts

Reply:
I would do this later.